src/VisualStudio/Core/Def/ProjectSystem/CPS/IWorkspaceProjectContextFactory.cs

```diff
@@ -42,5 +37,78 @@ internal interface IWorkspaceProjectContextFactory
             string? binOutputPath,
             string? assemblyName,
             CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Creates and initializes a new project and returns a <see
+        /// cref="IWorkspaceProjectContext"/> to lazily initialize the properties and items for the
+        /// project.  This method guarantees that either the project is added (and the returned task
+        /// completes) or cancellation is observed and no project is added.
+        /// </summary>
+        /// <param name="data">Providers access to msbuild evaluation data for the project.</param>
+        /// <param name="hierarchy">The IVsHierarchy for the project; this is used to track linked files across multiple projects when determining contexts.</param>
+        Task<IWorkspaceProjectContext> CreateProjectContextAsync(EvaluationData data, object? hierarchy, CancellationToken cancellationToken);
+
+        /// <summary>
+        /// Names of msbuild properties whose values that <see cref="CreateProjectContextAsync(EvaluationData, object?, CancellationToken)"/> will receive via <see cref="EvaluationData"/>.
+        /// </summary>
+        ImmutableArray<string> EvaluationPropertyNames { get; }
+
+        /// <summary>
+        /// Types of msbuild items whose values and metadata that <see cref="CreateProjectContextAsync(EvaluationData, object?, CancellationToken)"/> will receive via <see cref="EvaluationData"/>.
+        /// </summary>
+        ImmutableArray<string> EvaluationItemTypes { get; }
     }
+
+    internal abstract class EvaluationData
+    {
+        public abstract Guid ProjectGuid { get; }
+
+        /// <summary>
+        /// Unique across the entire solution for the life of the solution. 
+        /// Includes full path of the project, the GUID of project and the name of the config.
+        /// 
+        /// This will be unique across regardless of whether projects are added or renamed 
+        /// to match this project's original name. We include file path to make debugging easier.
+        /// 
+        /// For example:
+        ///      C:\Project\Project.csproj (Debug;AnyCPU {72B509BD-C502-4707-ADFD-E2D43867CF45})
+        ///      C:\Project\MultiTarget.csproj (Debug;AnyCPU;net45 {72B509BD-C502-4707-ADFD-E2D43867CF45})
+        /// </summary>
+        public abstract string ProjectUniqueName { get; }
+
+        public abstract string LanguageName { get; }
+
+        /// <summary>
+        /// Returns the value of property of the specified <paramref name="name"/>.
+        /// </summary>
+        /// <returns>
+        /// Returns null if the property is not set or the name is not listed in <see cref="IWorkspaceProjectContextFactory.EvaluationPropertyNames"/>.
+        /// </returns>
+        public abstract string? GetPropertyValue(string name);
```

Comment:
MSBuild APIs usually use empty strings to represent missing property values. If you want nulls here, then document that the return value will never be an empty value. 

Also any attempt to read a property the wasn't specified in `EvaluationPropertyNames` is probably an error and should throw.